### PR TITLE
372 nestedtypes

### PR DESCRIPTION
**Pull request by heshamMassoud**
#### Summary
Contains ProductType ReferenceResolution

#### Relevant Issues
#372 

#### Todo

- Tests
    - [x] Unit 
- [x] Documentation
- [x] Add Release Notes entry.

### CODE REVIEW

File: src/test/java/com/commercetools/sync/producttypes/helpers/ProductTypeReferenceResolverTest.java [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+            ProductTypeDraftBuilder.of("foo", "foo", "desc",
+                singletonList(attributeDefinitionDraft))
+                                   .build();
+
+        final ProductTypeSyncOptions syncOptions =
+            ProductTypeSyncOptionsBuilder.of(mock(SphereClient.class)).build();
+
+        final ProductTypeService productTypeService = mock(ProductTypeService.class);
+        when(productTypeService.fetchCachedProductTypeId(any()))
+            .thenReturn(CompletableFuture.completedFuture(Optional.of("foo")));
```
Comment (ahmetoz) [inline src/test/java/com/commercetools/sync/producttypes/helpers/ProductTypeReferenceResolverTest.java]: in mocks for any id, you are returning `foo` maybe return `x` like the prepared nested attribute?

File: src/main/java/com/commercetools/sync/producttypes/helpers/AttributeDefinitionReferenceResolver.java [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+                            attributeDefinitionDraft.getName());
+                    return exceptionallyCompletedFuture(
+                        new ReferenceResolutionException(errorMessage, exception.getCause()));
+                }
+            });
+
+    }
+
+    @Nonnull
+    private CompletionStage<AttributeDefinitionDraftBuilder> resolveNestedTypeReferenceOrSetOfNested(
```
Comment (ahmetoz) [inline src/main/java/com/commercetools/sync/producttypes/helpers/AttributeDefinitionReferenceResolver.java]: it looks like it's the main logic of the resolveReferences, maybe simply name it resolveReferences ?

File: src/main/java/com/commercetools/sync/producttypes/helpers/ProductTypeReferenceResolver.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+     *         a {@link ReferenceResolutionException}.
+     */
+    @Nonnull
+    public CompletionStage<ProductTypeDraft> resolveReferences(@Nonnull final ProductTypeDraft productTypeDraft) {
+        return resolveAttributeDefinitionsReference(ProductTypeDraftBuilder.of(productTypeDraft))
+            .thenApply(ProductTypeDraftBuilder::build);
+    }
+
+    @Nonnull
+    CompletionStage<ProductTypeDraftBuilder> resolveAttributeDefinitionsReference(
```
Comment (ahmetoz) [inline src/main/java/com/commercetools/sync/producttypes/helpers/ProductTypeReferenceResolver.java]:
```suggestion
    CompletionStage<ProductTypeDraftBuilder> resolveAttributeDefinitionReferences(
```

File: src/test/java/com/commercetools/sync/producttypes/helpers/ProductTypeReferenceResolverTest.java [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+
+        // test and assertion
+        assertThat(productTypeReferenceResolver.resolveReferences(productTypeDraft))
+            .isCompletedWithValue(expectedResolvedProductTypeDraft);
+    }
+
+    @Test
+    void resolveReferences_WithNoNestedTypeReferences_ShouldResolveReferences() {
+        // preparation
+
```
[the remark this answers is not in the document]
Comment (heshamMassoud) [reply src/test/java/com/commercetools/sync/producttypes/helpers/ProductTypeReferenceResolverTest.java]: 7227857

File: src/test/java/com/commercetools/sync/producttypes/helpers/ProductTypeReferenceResolverTest.java [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+import static java.util.Collections.singletonList;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+class ProductTypeReferenceResolverTest {
+
+    @Test
+    void resolveReferences_WithNullAttributes_ShouldResolveReferences() {
```
[the remark this answers is not in the document]
Comment (heshamMassoud) [reply src/test/java/com/commercetools/sync/producttypes/helpers/ProductTypeReferenceResolverTest.java]: ccd27c0

File: src/main/java/com/commercetools/sync/producttypes/helpers/AttributeDefinitionReferenceResolver.java (hunk 2 above)
Comment (heshamMassoud) [reply src/main/java/com/commercetools/sync/producttypes/helpers/AttributeDefinitionReferenceResolver.java]: 649694b

File: src/main/java/com/commercetools/sync/producttypes/helpers/ProductTypeReferenceResolver.java (hunk 3 above)
Comment (heshamMassoud) [reply src/main/java/com/commercetools/sync/producttypes/helpers/ProductTypeReferenceResolver.java]: b54c231

File: src/test/java/com/commercetools/sync/producttypes/helpers/ProductTypeReferenceResolverTest.java (hunk 1 above)
Comment (heshamMassoud) [reply src/test/java/com/commercetools/sync/producttypes/helpers/ProductTypeReferenceResolverTest.java]: but in the prepared nested attribute the `"x"` is a key actually. While this `"foo"` is an id.

File: src/test/java/com/commercetools/sync/producttypes/helpers/ProductTypeReferenceResolverTest.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (ahmetoz) [reply src/test/java/com/commercetools/sync/producttypes/helpers/ProductTypeReferenceResolverTest.java]: do not forget this one too: resolveReferences_WithNoAttributes_ShouldResolveReferences

[the remark this answers is not in the document]
Comment (heshamMassoud) [reply src/test/java/com/commercetools/sync/producttypes/helpers/ProductTypeReferenceResolverTest.java]: eacfd7a thank you :)

File: src/main/java/com/commercetools/sync/producttypes/helpers/AttributeDefinitionReferenceResolver.java [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+     * Given an {@link AttributeDefinitionDraft} this method attempts to resolve the ProductType references, which can
+     * be exist on attributeDefinition with an AttributeType: NestedType or SetType of NestedType, to return a
+     * {@link CompletionStage} which contains a new instance of the draft with the resolved references.
+     * The keys of the references are taken from the id field of the references.
+     *
+     * @param attributeDefinitionDraft the attributeDefinitionDraft to resolve its references.
+     * @return a {@link CompletionStage} that contains as a result a new attributeDefinitionDraft instance with resolved
+     *         references or if there is no productType existing with the given key the draft will be returned as is
+     *         without the reference resolved. In case an error occurs during reference resolution, a
+     *         {@link ReferenceResolutionException}.
```
[the remark this answers is not in the document]
Comment (heshamMassoud) [reply src/main/java/com/commercetools/sync/producttypes/helpers/AttributeDefinitionReferenceResolver.java]: c5df768

**Pull request merged**